src/rate_limiter.rs

```diff
@@ -20,27 +20,39 @@ use error::RoutingError;
 #[cfg(feature = "use-mock-crust")]
 use fake_clock::FakeClock as Instant;
 use itertools::Itertools;
+use lru_time_cache::LruCache;
 use maidsafe_utilities::serialisation::{self, SerialisationError};
 use messages::{MAX_PART_LEN, UserMessage};
 use sha3::Digest256;
 use std::cmp;
 use std::collections::BTreeMap;
 use std::mem;
 use std::net::IpAddr;
+use std::time::Duration;
 #[cfg(not(feature = "use-mock-crust"))]
 use std::time::Instant;
 use types::MessageId;
 
 /// The number of bytes per second the `RateLimiter` will "leak".
 const RATE: f64 = 5.0 * 1024.0 * 1024.0;
-/// The number of bytes per second of bandwidth distributed between all clients when demand is low.
+/// The minimum allowance (in bytes) for a single client at any given moment in the `RateLimiter`.
+/// This is slightly larger than `MAX_IMMUTABLE_DATA_SIZE_IN_BYTES` to allow for the extra bytes
+/// created by wrapping the chunk in a `UserMessage`, splitting it into parts and wrapping those in
+/// `RoutingMessage`s.
+const MIN_CLIENT_CAPACITY: u64 = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES + 10240;
+/// The maximum number of bytes the `RateLimiter` will "hold" at any given moment. This allowance
+/// is split equally between clients with entries in the `RateLimiter`. It is a soft limit in that
+/// it can be exceeded if there are enough client entries: each client will be allowed a
+/// hard-minimum of `MIN_CLIENT_CAPACITY` even if this means the `RateLimiter`'s total capacity
+/// exceeds the `SOFT_CAPACITY`.
 #[cfg(not(feature = "use-mock-crust"))]
 const SOFT_CAPACITY: u64 = 20 * 1024 * 1024;
-// Soft capacity has to be at least 2 * MIN_CLIENT_CAPACITY for the multi-client tests to work.
+/// For the mock-crust tests, we want a small `SOFT_CAPACITY` in order to trigger more rate-limited
+/// rejections. This must be at least `2 * MIN_CLIENT_CAPACITY` for the multi-client tests to work.
 #[cfg(feature = "use-mock-crust")]
 const SOFT_CAPACITY: u64 = 2 * MIN_CLIENT_CAPACITY;
-/// The minimum number of bytes per second of bandwidth supplied to any client.
-const MIN_CLIENT_CAPACITY: u64 = MAX_IMMUTABLE_DATA_SIZE_IN_BYTES + 10240;
+/// Duration for which entries are kept in the `overcharged` cache, in seconds.
+const OVERCHARGED_TIMEOUT_SECS: u64 = 3600;
```

Comment:
It might be enough, be we have plenty of room to play with here.  The entries are all pretty small and even at maximum throughput with just one client connected, the maximum number of entries in an hour would only be ~18,000.

Edit: @ustulation prefers 20secs too, so I'll go with the majority 😄